strawberryfields/backends/tfbackend/ops.py

```diff
@@ -159,44 +149,27 @@ def squeezed_vacuum_vector(r, theta, D, batched=False, eps=1e-32):
     return output
 
 
-def squeezer_matrix(r, theta, D, batched=False):
-    """creates the single mode squeeze matrix"""
-    r = tf.cast(r, tf.float64)
-    if not batched:
-        r = tf.expand_dims(r, 0)  # introduce artificial batch dimension
-    r = tf.reshape(r, [-1, 1, 1, 1])
-    theta = tf.cast(theta, def_type)
-    theta = tf.reshape(theta, [-1, 1, 1, 1])
+@tf.custom_gradient
+def single_squeezing_matrix(r, phi, D, dtype=def_type.as_numpy_dtype):
+    """creates a single mode squeezing matrix"""
+    r = r.numpy()
+    phi = phi.numpy()
+    gate = squeezing_tw(r, phi, D, dtype)
 
-    rng = np.arange(D)
-    n = np.reshape(rng, [-1, D, 1, 1])
-    m = np.reshape(rng, [-1, 1, D, 1])
-    k = np.reshape(rng, [-1, 1, 1, D])
-
-    phase = tf.exp(1j * theta * (n - m) / 2)
-    signs = squeeze_parity(D).reshape([1, D, 1, D])
-    mask = np.logical_and(
-        (m + n) % 2 == 0, k <= np.minimum(m, n)
-    )  # kills off terms where the sum index k goes past min(m,n)
-    k_terms = (
-        signs
-        * tf.pow(tf.sinh(r) / 2, mask * (n + m - 2 * k) / 2)
-        * mask
-        / tf.pow(tf.cosh(r), (n + m + 1) / 2)
-        * tf.exp(
-            0.5 * tf.math.lgamma(tf.cast(m + 1, tf.float64))
-            + 0.5 * tf.math.lgamma(tf.cast(n + 1, tf.float64))
-            - tf.math.lgamma(tf.cast(k + 1, tf.float64))
-            - tf.math.lgamma(tf.cast((m - k) / 2 + 1, tf.float64))
-            - tf.math.lgamma(tf.cast((n - k) / 2 + 1, tf.float64))
-        )
-    )
-    output = tf.reduce_sum(phase * tf.cast(k_terms, def_type), axis=-1)
+    def grad(dy):
+        Dr, Dphi = grad_squeezing_tw(gate, r, phi)
+        grad_r = 2 * tf.math.real(tf.reduce_sum(dy * tf.math.conj(Dr)))
+        grad_phi = 2 * tf.math.real(tf.reduce_sum(dy * tf.math.conj(Dphi)))
+        return grad_r, grad_phi, None
 
-    if not batched:
-        # remove extra batch dimension
-        output = tf.squeeze(output, 0)
-    return output
+    return gate, grad
```

Comment:
Might be missing a piece, it seems to me that this was not a tuple previously :thinking:

Reply:
Note that this function is wrapped in the `@tf.custom_gradient decorator`. It now needs to return a tuple with the value of the output and a function (`grad(dy)`) that applies the vector-Jacobian multiplication to propagate the upstream gradient.